apps/passport-server/src/services/frogcryptoService.ts

```diff
@@ -0,0 +1,272 @@
+import { logger } from "../util/logger";
+import { RollbarService } from "./rollbarService";
+
+export class FrogcryptoService {
+  private readonly context: ApplicationContext;
+  private readonly rollbarService: RollbarService | null;
+  private readonly verificationPromiseCache: LRUCache<
+    string,
+    Promise<string | null>
+  >;
+
+  public constructor(
+    context: ApplicationContext,
+    rollbarService: RollbarService | null
+  ) {
+    this.context = context;
+    this.rollbarService = rollbarService;
+    this.verificationPromiseCache = new LRUCache<
+      string,
+      Promise<string | null>
+    >({
+      max: 1000
+    });
+  }
+
+  public async getFeeds(): Promise<FrogCryptoFeedConfig[]> {
+    return FROGCRYPTO_FEEDS;
+  }
+
+  public async getUserState(
+    req: FrogCryptoUserStateRequest
+  ): Promise<FrogCryptoUserStateResponseValue> {
+    const semaphoreId = await this.cachedVerifySignaturePCD(req.pcd);
+    if (!semaphoreId) {
+      throw new PCDHTTPError(400, "invalid PCD");
+    }
+
+    const userFeeds = await fetchUserFeedsState(
+      this.context.dbPool,
+      semaphoreId
+    );
+
+    const allFeeds = _.keyBy(await this.getFeeds(), "id");
+
+    return {
+      feeds: userFeeds.map((userFeed) =>
+        this.computeUserFeedState(userFeed, allFeeds[userFeed.feed_id])
+      )
+    };
+  }
+
+  public async reserveFrogData(
+    pcd: SerializedPCD<SemaphoreSignaturePCD>,
+    feed: FrogCryptoFeedConfig
+  ): Promise<IFrogData> {
+    const semaphoreId = await this.cachedVerifySignaturePCD(pcd);
+    if (!semaphoreId) {
+      throw new PCDHTTPError(400, "invalid PCD");
+    }
+
+    await initializeUserFeedState(this.context.dbPool, semaphoreId, feed.id);
+
+    return sqlTransaction(
+      this.context.dbPool,
+      "reserve frog",
+      async (client) => {
+        const lastFetchedAt = await updateUserFeedState(
+          client,
+          semaphoreId,
+          feed.id
+        ).catch((e) => {
+          if (e.message.includes("could not obtain lock")) {
+            throw new PCDHTTPError(
+              429,
+              "There is another frog request in flight!"
+            );
+          }
+          throw e;
+        });
+        if (!lastFetchedAt) {
+          const e = new Error("User feed state unexpectedly not found!");
+          logger(`Error encountered while serving feed:`, e);
+          this.rollbarService?.reportError(e);
+          throw e;
+        }
+
+        const { nextFetchAt } = this.computeUserFeedState(
+          {
+            feed_id: feed.id,
+            last_fetched_at: lastFetchedAt
+          },
+          feed
+        );
+        if (nextFetchAt > Date.now()) {
+          throw new PCDHTTPError(403, `Next fetch available at ${nextFetchAt}`);
+        }
+
+        const frogData = await sampleFrogData(this.context.dbPool, feed.biomes);
+        if (!frogData) {
+          throw new PCDHTTPError(404, "Frog Not Found");
+        }
```

Comment:
this is funny, but idk if 404 is appropriate here. could be funny enough that it's fine.

Reply:
yes yes. happy to revise if it ended up a case we need to explicitly handle